File: Cpp/DataStructuresAndAlgos/Heap/minheap.cpp
#include <iostream>
#include <cmath>

void print(int* &array, int size){
    std::cout << "\nPrinting array: " << std::endl;
    for (int i = 0; i < size; ++i){
        std::cout << array[i] << " ";
    }
    std::cout << std::endl;
}

int getRightChild(int index) {
    int right_c_index = 2*index + 2; 
    return right_c_index;
}

int getLeftChild(int index) {
    int left_c_index = 2*index + 1; 
    return left_c_index;
}

int getParent(int index) {
    int parent_index = (index+1)/2 -1;
    return parent_index;
}

int getMinValForPair(int index, int* &array) {
    // Gets the minimum value between two adjacent heap elements that
    // have the same parent.
    if (index % 2 == 0) {
        return std::min(array[index], array[index-1]);
    } else {
        return std::min(array[index], array[index+1]);
    }
}

int getHeight(int index){
    return static_cast<int>(log2(static_cast<float>(index)));
}

int indexOfSmallerChild(int index_a, int index_b, int* &array) {
    if (array[index_a] < array[index_b]){
        return index_a;
    } else {
        return index_b;
    }
}

int pop(int index, int* &array, int size){
    // We start at the point of deletion (top), and bubble values up.
    int value = array[index];

    int parent_index = index;
    int child_index = index;
    while (child_index / 2 < size) {
        child_index = indexOfSmallerChild(getLeftChild(parent_index), getRightChild(parent_index), array);
        if (child_index > size-1){
            array[parent_index] = size*2;
        } else {
            array[parent_index] = array[child_index];
        }
        parent_index = child_index;
    }
    return value;
}

void insertion(int value, int* &array, int length){
    int insert_index =  length-1;
    int heap_height = getHeight(length);

    // We keep going up until we find a value the inserted value is larger than (or we get to the top)
    while ((value < getMinValForPair(insert_index, array)) && (insert_index > 0)){
        std::cout << "Value : " << value << "; smaller than: " << getMinValForPair(insert_index, array) << std::endl;
        insert_index = getParent(insert_index);
        std::cout << "Next Insert index: " << insert_index << std::endl;
    }

    // We insert the insertion value, and store the old value.
    int temp = array[insert_index]; // store the old value
    array[insert_index] = value; // insert new value



    // We swap the values back down to the bottom of the heap 
    int new_index = insert_index;
    int temp2 = 0;
    for (int i = getHeight(insert_index); i < heap_height; ++i ) {
        print(array, length);
        std::cout << "New index " << new_index << std::endl;
        new_index = getLeftChild(new_index); // index we'll take the place of
        if (new_index > (length-1)) {
            std::cout << "Exceeding array bounds" << std::endl;
            break;
        }
        std::cout << "New index " << new_index << std::endl;
        temp2 = array[new_index]; // value we'll need to move next iteration
        array[new_index] = temp; // we take the place of the value
        std::cout << temp << " took the place of " << temp2 << std::endl;
        temp = temp2;
    }
}

int main() {

    int size = 15;
    int array[size];
    // We leave one value of the array free for insertion
    for (int i = 0; i < size-1; ++i){
        array[i] = i;
    }
    array[size-1] = 1000;

    int* pArray = &array[0];
    print(pArray, size);


    pop(0, pArray, size);
    print(pArray, size);

    insertion(0, pArray, size-1);
    print(pArray, size);

    return 0;
}

File: Cpp/DataStructuresAndAlgos/Heap/minheap.py
# Python implementation, just to practice algos in both languages
import pdb
import numpy as np
from ipdb_hook import ipdb_sys_excepthook
ipdb_sys_excepthook()

class MinHeap:

    def __init__(self, array):
        assert len(array.shape), "Array should be 1D"
        self.heap = array

    def print(self):
        print(f"Heap: {self.heap}")

    def getParent(self, index):
        return int(np.floor((index+1)/2) -1)

    def getLeftChild(self, index):
        return 2*index +1
    
    def getRightChild(self, index):
        return 2*index +2
    
    def getMinChildIndex(self, index: int) -> int:
        """
        Returns the index of the smaller of the two children of a value given by an index.
        """

        left_child = self.heap[self.getLeftChild(index)]
        right_child = self.heap[self.getRightChild(index)]
        if left_child <= right_child:
            return self.getLeftChild(index)
        else:
            return self.getRightChild(index)

    def validIndex (self, index):
        return index < len(self.heap)-1

    def heapifyUp(self, value: int, index: int) -> int:
        """
        Moves a value up until the heap is correct (all parents are smaller or equal to children).
        Returns the index the value should be inserted at.
        """
        parent_index = self.getParent(index)
        if value < self.heap[parent_index]:
            index = self.heapifyUp(value, parent_index)
        
        return index

    def heapifyDown(self, value, index):
        """
        Moves a value down until the heap is correct. This takes `value` at `index`,
        and moves it down by comparing it to the smaller of its two children. If it is 
        smaller, then it swaps places with that child.

        Return None.
        """

        print(f"Dealing with value {value} at index {index}")

        # If we don't have valid children, we stop
        if not self.validIndex(self.getRightChild(index)):
            # Zeroes below, NaNs?
            print(f"No place, creating a new entry in heap")
            extra = np.ones((1, self.getRightChild(index)+2 - len(self.heap))) *100
            self.heap = np.append(self.heap,extra)
            self.heap[self.getRightChild(index)] = value
            # I may have to push to the array here (append, at a specific index)
            return

        down_index = self.getMinChildIndex(index)
        down_val = self.heap[down_index]
        # This logic shouldn't be conditional, it should force the swap.
        # Value will awlays be smaller than down_val, by the nature of a minheap.
        self.heap[down_index] = value
        print(f"{value} took the spot of {down_val}: \n{self.heap}")
        self.heapifyDown(down_val, down_index)


    def insert(self, value):
        # Add to the bottom, at any value is fine, and then heapify up.
        # Once we've inserted it, the "replaced" value, will heapify down.

        insertion_index = len(self.heap)
        print(f"Length: {len(self.heap)}")
        insertion_index = self.heapifyUp(value, insertion_index)

        print(f"We will insert {value} at index {insertion_index}")

        # We insert the new value
        temp = self.heap[insertion_index] 
        self.heap[insertion_index] = value

        # We heapify the replaced value down
        self.heapifyDown(temp, insertion_index)


    def pop(self):
        # Delete (from top of heap), and then heapify down.
        index = 0

        # Current value we are replacing
        value = self.heap[index]

        while True:
            # We get the smaller child of the current value, which will replace it.
            # We check the child is valid, if not, the current value is already at the bottom
            if not self.validIndex(self.getRightChild(index)):
                self.heap[index] = 100
                break
            
            down_index = self.getMinChildIndex(index)

            # If the child is valid, we move it up
            print(f"Moving {self.heap[down_index]} to the spot formerly occupied by {self.heap[index]}")
            down_val = self.heap[down_index]
            self.heap[index] = down_val
            self.print()
            index = down_index

        #self.length -=1
        return value



array = np.arange(15)
minheap = MinHeap(array) 
#import pdb
#pdb.set_trace()
print(minheap.heap)


popped = minheap.pop()
print(f"Popped value: {popped}; Heap: {minheap.heap}")



value = 6
minheap.insert(value)
print(f"Inserted {value}; Heap: {minheap.heap}")

value = 2
minheap.insert(value)
print(f"Inserted {value}; Heap: {minheap.heap}")